### ISSUE TITLE
[Feature]: Async cudaHostRegister in SimpleCPUOffloadConnector to unblock  startup

### ISSUE THREAD

**Issue by wenduo** [vllm-project/vllm#42632]
## Motivation

`KVConnectorBase_V1.register_kv_caches()` is called synchronously from the worker init path (`vllm/v1/worker/gpu_model_runner.py:7041`, inside `initialize_kv_cache`). For `SimpleCPUOffloadConnector`, the worker-side handler issues a single `cudaHostRegister` over the entire CPU KV region:

```python
# vllm/v1/simple_kv_offload/worker.py:165-167
tensor = torch.zeros(cpu_shape, dtype=gpu_tensor.dtype, device="cpu")
if pin_memory:
    pin_tensor(tensor)   # -> cudaHostRegister, vllm/v1/simple_kv_offload/cuda_mem_ops.py:24
```

**At TB-scale this is operationally fatal.** Modern offload deployments are sizing `cpu_bytes_to_use` into the multi-TB range to maximize prefix reuse. `cudaHostRegister` over that much memory is dominated by the kernel pinning every page (NVIDIA driver lock + RDMA cgroup accounting + page-table walks) and we have observed this single call running for many minutes on commodity hardware.

For the entire pin window:

- `register_kv_caches()` does not return.
- Worker `initialize_cache` collective_rpc does not return → engine startup does not finish.
- The HTTP server cannot accept *any* request, even short prompts that would never touch the offload tier.
- `/health` stays un-ready; autoscaler readiness probes time out.

**This breaks elastic scaling under bursty traffic.** When the autoscaler brings new vLLM replicas up in response to a spike, those replicas sink the entire pin window before serving a single request. The burst arrives, the replicas are still pinning, and traffic goes unserved. Slow-start probes don't help — the engine has literally not finished booting.

The fix is to let the engine become serve-ready first and pin in the background. While `cudaHostRegister` runs, requests are served on the GPU-only path; once pinning completes, the connector silently joins in and prefix hits from CPU resume.

The same blocking pattern exists in `OffloadingConnector` (CPU spec, `vllm/v1/kv_offload/cpu/gpu_worker.py:89`, `pin_mmap_region`) and is a candidate for a follow-up PR.

## Proposal

Make `SimpleCPUOffloadConnector`'s host-memory pin asynchronous, with the scheduler gating connector traffic until every TP rank reports "register done". Default-off via an `extra_config` flag, behavior-preserving.

**Worker side** (`vllm/v1/simple_kv_offload/worker.py`):

- When `async_register_cache` is set, `register_kv_caches()` allocates CPU tensors on the main thread (no pin work) and kicks off a `threading.Thread` that does:
  1. `pin_tensor(...)` per tensor (the long part).
  2. `_backend.init(...)` (wires `load_stream` / `store_stream` and the copy backend).
  3. Flips `self._ready = True`.
- Main thread returns immediately so worker init completes and the engine becomes serve-ready.

**Scheduler side** (`vllm/v1/simple_kv_offload/manager.py`):

- `SimpleCPUOffloadScheduler` gains `_ready: bool` (initialized to `not async_register_cache`) and `_ready_ranks: set[int]`.
- `get_num_new_matched_tokens(...)` short-circuits to `(0, False)` while `not self._ready` — requests are served, just without prefix hits from CPU offload, until the gate opens.
- `build_connector_meta(...)` returns an empty `SimpleCPUOffloadMetadata` while `not ready` so the worker never tries to load/store before the backend is initialized.

**Ready signal** (`SimpleCPUOffloadWorkerMetadata`):

- Adds `ready_tprank: int | None = None`. Worker's `build_connector_worker_meta()` sets it once when `self._ready` flips True. `update_connector_output(...)` accumulates `ready_tprank` across ranks; when the set size hits `_expected_worker_count`, scheduler flips `self._ready = True`. Re-uses the existing worker→scheduler channel — no new RPC.

**Config**:

- New `extra_config["async_register_cache"] = bool` (default `False`). Documented in the connector docstring as recommended once `cpu_bytes_to_use_per_rank` enters the multi-hundred-GB / TB range.

## Failure modes considered

- If the background pin thread fails, worker logs `logger.exception(...)` and the gate never opens → safe degradation to GPU-only serving (no CPU offload).
- Crash inside the thread can't reach `transfer_async`: the scheduler-side gate prevents transfer calls from being built at all while `not ready`.
- Race with `bind_connector_metadata` during the gap: `build_connector_meta` produces empty metadata while `not ready`, so there is nothing for the worker to bind.

## Scope of the proposed first PR

Only `SimpleCPUOffloadConnector`. Follow-ups (separate PRs, after this lands and the pattern is reviewed):

- `OffloadingConnector` with CPU spec: same pattern around `pin_mmap_region`.
- `NixlConnector` / `MooncakeConnector`: their registration goes through third-party libs (NIXL `register_memory`, Mooncake `batch_register_memory`); handshake-vs-register ordering needs separate discussion.

## Test plan

- Unit: `SimpleCPUOffloadScheduler` returns `(0, False)` while not ready; flips after `tp_size` `ready_tprank` reports.
- Unit: `SimpleCPUOffloadWorker.register_kv_caches(async_register_cache=True)` returns in milliseconds; `self._ready` flips after the thread joins.
- Integration: launch with `cpu_bytes_to_use=200GB, async_register_cache=true`; mock `cudaHostRegister` with a sleep, confirm `/health` becomes ready during the pin and a request submitted in the gap completes (without prefix hit from CPU offload).
- Regression: with `async_register_cache=false` (default), behavior identical to today.

**Comment by YannikHinteregger** [comment]
Found myself with the same issue.
Opened a PR for this #44127

**Comment by wenduo** [comment]
there are some trouble in my scence.  Parallel cudamemoryRegister will have serious performence influence for parallel request. There are 3 ways to handle:
1. register a  small bouncing buffer instead
2. hugetlbfs
3. some sm-free way to swapin/swapout.
All have its goods and bads.  No silver bullet now. 
So I wonder how vllm community see this issue.  I see  the bouncing buffer is the best way now.

**Comment by YannikHinteregger** [comment]
I see, makes sense. Just double checking that I read your message right: My PR implements the design from your initial post. You're saying this approach won't hold at TB scale and will degrade token throughput during pinning, right?

I'd suggest merging my PR anyway and treating it as a first iteration. It's behind an opt-in config flag and very useful for smaller nodes (few hundred GB of RAM). Low risk, net positive change.

For a follow-up / second iteration: what do you think about incremental + adaptive pinning?
1. Pin in small chunks (e.g. ~1 GB) so we're not holding the heavy lock for long in the background thread. That should keep token throughput high
2. Then pin on demand, keep ~20% of pinned memory free as a buffer, pin more in the background when needed

Has this been considered already?

**Comment by YannikHinteregger** [comment]
Spent a while on this one, writing up what I found so nobody repeats it.

@ivanium reviewed my async pinning PR (#44127) and thought the threading was too risky and that this needs a bigger effort rather than a fix bolted onto one connector. Fair. So instead of hiding the cost I tried to make it smaller.

All dead ends. Benchmarks and logs: [YannikHinteregger/vllm-cpu-pin-benchmarks](https://github.com/YannikHinteregger/vllm-cpu-pin-benchmarks)

**Async**: Declined, too complex for now

**Chunking the register calls**: No difference inside one allocation. Splitting into 80 separate allocations, which is what the connector does per layer, costs about 1.9x. That works out to ~10 GiB/s, so ~1.7 min per TB.

**Threads**: No difference, but my test was broken. Torch doesn't release the GIL for `cudaHostRegister`, so the threads never ran in parallel. Can't say anything either way on this one.

**Bigger pages**: works, 2 MB pages instead of the usual 4 KB gives roughly 1.5x on allocation plus pinning:

```
  size   base t1   huge t1   x(t1) p1   x(t1) p2   THP GiB
    8G      9.32      4.35      2.14x      2.69x       8.0
   32G     20.13     19.01      1.06x      1.44x      31.9
   64G     31.07     20.73      1.50x      1.46x      63.9
  128G     51.02     33.81      1.51x      1.62x     128.0
  192G     70.01     45.30      1.55x      1.72x     191.9
```

But PyTorch already does this if you set `THP_MEM_ALLOC_ENABLE=1`. Same pages,
same speedup, my own allocator on top added nothing:

```
  size   zeros t1   custom t1    x(t1)   zeros THP   custom THP
   32G      15.66       14.84    1.06x        31.8         31.9
  192G      47.45       50.00    0.95x       191.8        191.9
```

So nothing to build here.

If pinning is slow for you, try that env var and measure. Expect about a third off. Two catches: it needs contiguous free memory and on a fragmented host it can make startup much slower instead — I hit that hard on a box that had been churning big allocations for hours. It's also process-wide, so it affects every large CPU allocation, not just the KV tier.

None of this closes the issue though. Even with huge pages, 192 GB still blocks startup for ~45 s. Making the pin cheaper isn't enough. Getting it off the critical path is the actual fix and that's the part worth doing properly rather than per connector.

Leaving this one for now. Happy to help and contribute if someone picks it up again.
